### PR TITLE
Improve Travis CI setup and fix errors/warnings to run with modern compilers

### PR DESCRIPTION
**Pull request by AndyMender**
Unfortunately, this PR went a little out of hand when I was working on it so it contains some needless code style changes. I will fix them so they don't generate excess diff lines.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -59,7 +59,7 @@ GLuint CreateTexture(char const* Filename)
 
 | Service | System | Compiler | Status |
 | ------- | ------ | -------- | ------ |
-| [Travis CI](https://travis-ci.org/g-truc/gli)| Linux 64 bits | Clang 3.4, Clang 3.8, GCC 4.9 | [![Travis CI](https://travis-ci.org/g-truc/gli.svg?branch=master)](https://travis-ci.org/g-truc/gli)
+| [Travis CI](https://travis-ci.org/g-truc/gli)| Linux 64 bits | Clang 7.0, Clang 9.0, GCC 7.4, GCC 9, GCC 10 | [![Travis CI](https://travis-ci.org/g-truc/gli.svg?branch=master)](https://travis-ci.org/g-truc/gli)
```
Comment (AndyMender) [inline readme.md:62]: The mentioned extra compiler compatibility. The `.travis.yml` config added to the PR provides configurations for these compilers.

File: test/core/core_flip.cpp [hunk 2]
```diff
@@ -31,48 +31,45 @@ int main()
...
-		glm::f32vec4(1.0, 0.5, 0.0, 1.0), glm::f32vec4(0.0, 0.5, 1.0, 1.0));
+		glm::f32vec4(1.0, 0.5, 0.0, 1.0),
+		glm::f32vec4(0.0, 0.5, 1.0, 1.0));
 
-	Error += test_texture(
-		gli::texture2d(gli::FORMAT_RGB_DXT1_UNORM_BLOCK8, TextureSize, Levels),
-		gli::detail::dxt1_block{63721, 255, 228, 144, 64, 0},
-		gli::detail::dxt1_block{2516, 215, 152, 173, 215, 106});
+	Error += test_texture(gli::texture2d(gli::FORMAT_RGB_DXT1_UNORM_BLOCK8, TextureSize, Levels),
+		gli::detail::dxt1_block{ 63721, 255, { 228, 144, 64, 0 } },
```
Comment (AndyMender) [inline test/core/core_flip.cpp:59]: Below are not code style fixes, The `gli::detail::dxt*_block` structs receive a different number number of arguments, but previous compilers would ignore this.

File: .travis.yml [hunk 3]
```diff
@@ -1,37 +1,86 @@
...
+    - compiler: clang-10
+    - name: "Build with LLVM/Clang 10"
       addons:
         apt:
           sources:
-            - ubuntu-toolchain-r-test
+          - sourceline: 'deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic-10 main'
+          key_url: 'https://apt.llvm.org/llvm-snapshot.gpg.key'
           packages:
-            - g++-4.9
```
Comment (AndyMender) [inline .travis.yml:13]: I did remove the previous compiler build entries, but if they're still needed, I can re-add them.

File: test/core/generate_mipmaps_sampler2d.cpp [hunk 4]
```diff
@@ -7,40 +13,60 @@
...
-		if(Texture.levels() > 1)
-			Error += LoadC == Black ? 0 : 1;
+		// If texture has multiple mipmap levels, texel from top level must match base color
+		if (Texture.levels() > 1)
+			Error += (TopTexel == BaseColor) ? 0 : 1;
 
+		// create a texture view and a sampler
 		gli::texture2d TextureView(gli::view(Texture, 0, 0));
 		gli::fsampler2D SamplerA(gli::texture2d(gli::duplicate(Texture)), gli::WRAP_CLAMP_TO_EDGE);
-		SamplerA.generate_mipmaps(gli::FILTER_LINEAR);
```
Comment (AndyMender) [inline test/core/generate_mipmaps_sampler2d.cpp:24]: `gli::FILTER_LINEAR` fails some of the test iterations. It was replaced with the value of `Filter` passed to the `test` function.

**Comment by AndyMender** [comment]
I removed most of the clean-up lines, but still left in changes like `int Error = 0` instead of `int Error(0)` and extra parentheses to make the `Error` logic operations in tests clearer. I can remove these as well if you consider them unneeded :).

File: test/core/core_texture.cpp [hunk 5]
```diff
@@ -6,22 +10,16 @@
...
-		int Error(0);
+		int Error = 0;
 
-		std::vector<int> Sizes;
-		Sizes.push_back(16);
-		Sizes.push_back(32);
-		Sizes.push_back(15);
-		Sizes.push_back(17);
-		Sizes.push_back(1);
+		std::vector<int> Sizes = { 16, 32, 15, 17, 1 };
```
Comment (AndyMender) [inline test/core/core_texture.cpp:22]: Debatable whether changes like these are needed. Theoretically, initializing the vector only once avoids the needless bookkeeping on `push_back` calls.

File: test/core/core_texture.cpp [hunk 6]
```diff
@@ -372,7 +371,7 @@ namespace perf_generic_creation
 		}
 
 		std::clock_t TimeEnd = std::clock();
-		printf("Generic texture creation performance test: %d\n", TimeEnd - TimeBegin);
```
Comment (AndyMender) [inline test/core/core_texture.cpp:375]: A matter of opinion, but `std::cout` is a little more flexible when converting data to text and `std::endl` adds platform-specific newline symbols.

File: test/core/test_size.cpp [hunk 7]
```diff
@@ -117,83 +117,85 @@ namespace can_compute_npot_texture_size
...
-			Error += Texture.size() == gli::block_size(gli::FORMAT_RGB_DXT1_UNORM_BLOCK8) ? 0 : 1;
+			Error += (Texture.size() == gli::block_size(gli::FORMAT_RGB_DXT1_UNORM_BLOCK8)) ? 0 : 1;
 		}
 
 		{
 			gli::texture2d Texture(gli::FORMAT_RGB_DXT1_UNORM_BLOCK8, gli::texture2d::extent_type(9, 5), 1);
 
-			gli::texture2d::size_type const CurrentSize = Texture.size();
-			gli::texture2d::size_type const ExpectedSize = gli::block_size(gli::FORMAT_RGB_DXT1_UNORM_BLOCK8) * 3 * 2;
+			const gli::texture2d::size_type CurrentSize = Texture.size();
```
Comment (christophe-lunarg) [inline test/core/test_size.cpp:195]: coding style use "const" after the type

**Review by christophe-lunarg — CHANGES_REQUESTED** [review]
It's a bit unfortunate that so many changes obfuscate the actual work.

This said, it's more the compiler coverage that I am concerned about. It would be great to keep compatibility with old compilers.

**Comment by christophe-lunarg** [comment]
These Travis C.I. changes are very welcome !

**Comment by AndyMender** [comment]
I decided that it makes a little more sense to focus on the Travis CI setup alone and not mud it with additional commits. I opened a new PR here: https://github.com/g-truc/gli/pull/183

Fixes for compiler warnings can be tackled separately, I think.

**Pull request closed without merging**